hypothesis-python/tests/array_api/test_indices.py

```diff
@@ -0,0 +1,122 @@
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis/
+#
+# Most of this work is copyright (C) 2013-2021 David R. MacIver
+# (david@drmaciver.com), but it contains contributions by others. See
+# CONTRIBUTING.rst for a full list of people who may hold copyright, and
+# consult the git log if you need to determine who owns an individual
+# contribution.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+#
+# END HEADER
+
+import math
+
+import pytest
+
+from hypothesis import assume, given, strategies as st
+
+from tests.array_api.common import xp, xps
+from tests.common.debug import find_any
+
+pytestmark = [pytest.mark.mockable_xp]
+
+
+@pytest.mark.parametrize(
+    "condition",
+    [
+        lambda ix: Ellipsis in ix,
+        lambda ix: Ellipsis not in ix,
+    ],
+)
+def test_indices_options(condition):
+    indexers = (
+        xps.array_shapes(min_dims=1, max_dims=32)
+        .flatmap(lambda shape: xps.indices(shape))
+        .map(lambda idx: idx if isinstance(idx, tuple) else (idx,))
+    )
+    find_any(indexers, condition)
+
+
+def test_indices_can_generate_empty_tuple():
+    find_any(xps.indices(shape=(0, 0), allow_ellipsis=True), lambda ix: ix == ())
+
+
+def test_indices_can_generate_non_tuples():
+    find_any(
+        xps.indices(shape=(0, 0), allow_ellipsis=True),
+        lambda ix: not isinstance(ix, tuple),
+    )
+
+
+def test_indices_can_generate_long_ellipsis():
+    # Runs of slice(None) - such as [0,:,:,:,0] - can be replaced by e.g. [0,...,0]
+    find_any(
+        xps.indices(shape=(1, 0, 0, 0, 1), max_dims=3, allow_ellipsis=True),
+        lambda ix: len(ix) == 3 and ix[1] == Ellipsis,
+    )
+
+
+@given(
+    xps.indices(shape=(0, 0, 0, 0, 0), max_dims=5).filter(
+        lambda idx: isinstance(idx, tuple) and Ellipsis in idx
+    )
+)
+def test_indices_replaces_whole_axis_slices_with_ellipsis(idx):
+    # When a shape's dimension sizes are 0, ... in indices replaces all slice
+    # and integer entries in the index.
```

Comment:
```suggestion
    # `slice(None)` (aka `:`) is the only valid index for an axis of size
    # zero, so if all dimensions are 0 then a `...` will replace all the
    # slices because we generate `...` for entire contiguous runs of `:`
```

Ooh, this is surprisingly tricky!  Fixed comment explains why there's no integer indices.

Reply:
Comitted this suggestion for both `tests/array_apy` and `tests/numpy` instants of this test.

Might be a good idea to have me refactor so that these tests covering solely stuff in `_array_helpers.py` are in say `tests/array_helpers`.

Reply:
Let's leave any further work on that for a future PR, after this one is merged 😁